text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+
+**`GetCorrelation(context, key) -> value`**  
+To access the value for a label set by a prior event, the Correlations API 
+provides a function which takes a context and a key as input, and returns a value.
+
+**`SetCorrelation(context, key, value) -> context`**  
+To record the value for a label, the Correlations API provides a function which 
+takes a context, a key, and a value as input, and returns an updated context 
+which contains the new value.
+
+**`RemoveCorrelation(context, key) -> context`**  
+To delete a label, the Correlations API provides a function 
+which takes a context and a key as input, and returns an updated context which 
+no longer contains the selected key-value pair.
+
+**`ClearCorrelations(context) -> context`**  
+To avoid sending any labels to an untrusted process, the Correlation API 
+provides a function to remove all Correlations from a context.
+
+**`GetCorrelationPropagator() -> (HTTP_Extractor, HTTP_Injector)`**  
+To deserialize the previous labels set by prior processes, and to serialize the 
+current total set of labels and send them to the next process, the Correlations 
+API provides a function which returns a Correlation-specific implementation of 
+the `HTTPExtract` and `HTTPInject` functions found in the Propagation API.
+
+# Context Propagation
+
+## Context API
+
+Cross-cutting concerns access data in-process using the same, shared context 
+object. Each concern uses its own namespaced set of keys in the context, 
+containing all of the data for that cross-cutting concern.
+
+**`GetValue(context, key) -> value`**  
+To access the local state of an concern, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+**`SetValue(context, key, value) -> context`**  
+To record the local state of a cross-cutting concern, the Context API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**`GetCurrent() -> context`**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+**`SetCurrent(context)`**  
+To associate a context with program execution, the Context API provides a 
+function which takes a Context.
+
+
+
+## Propagation API
+
+Cross-cutting concerns send their state to the next process via propagators: 
+functions which read and write context into RPC requests. Each concern creates a 
+set of propagators for every type of supported medium - currently only HTTP 
+requests.
+
+**`Extract(context, []http_extractor, headers) -> context`**  
```

Comment:
It may be easier to implement it with lambdas using generics. Like it is done in C#: https://github.com/open-telemetry/opentelemetry-dotnet/blob/605760cf5ae0511b94d475a754f5ca1a5ec35f41/src/OpenTelemetry.Api/Context/Propagation/ITextFormat.cs#L35-L51

Reply:
I think this is a language-specific recommendation, and finer detail than the OTEP (or the spec) would get into.